code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -69,27 +71,62 @@ public Account getAdminAccount() {
     }
 
     @Override
-    public List<Account> searchAccounts(String name) {
-        List<Account> accounts = new ArrayList<>();
-        accounts.addAll(create()
-                .selectFrom(ACCOUNT)
-                .where(ACCOUNT.STATE.eq(CommonStatesConstants.ACTIVE)
-                        .and(ACCOUNT.NAME.contains(name))
-                        .and(ACCOUNT.KIND.eq(AccountConstants.USER_KIND)
-                                .or(ACCOUNT.KIND.eq(AccountConstants.ADMIN_KIND))))
-                .orderBy(ACCOUNT.ID.asc()).fetch());
-        return accounts;
+    public List<Account> searchUsers(String username) {
+        try {
+            List<Account> accounts = new ArrayList<>();
+            accounts.addAll(create()
+                    .select(ACCOUNT.fields())
+                    .from(ACCOUNT)
+                    .join(CREDENTIAL)
+                    .on(CREDENTIAL.ACCOUNT_ID.eq(ACCOUNT.ID))
+                    .where(ACCOUNT.STATE.eq(CommonStatesConstants.ACTIVE)
+                            .and(CREDENTIAL.STATE.eq(CommonStatesConstants.ACTIVE))
+                            .and(CREDENTIAL.PUBLIC_VALUE.contains(username))
+                            .and(CREDENTIAL.KIND.eq("password")))
+                    .orderBy(ACCOUNT.ID.asc()).fetchInto(AccountRecord.class));
+            return accounts;
+        } catch (InvalidResultException e) {
+            throw new ClientVisibleException(ResponseCodes.CONFLICT, "MultipleSameUsernames");
+        }
     }
 
     @Override
-    public Account getByName(String name) {
-        return create()
-                .selectFrom(ACCOUNT)
-                .where(ACCOUNT.STATE.eq(CommonStatesConstants.ACTIVE)
-                        .and(ACCOUNT.NAME.eq(name))
-                        .and(ACCOUNT.KIND.eq(AccountConstants.USER_KIND)
-                                .or(ACCOUNT.KIND.eq(AccountConstants.ADMIN_KIND))))
-                .orderBy(ACCOUNT.ID.asc()).fetchOneInto(Account.class);
+    public Account getByUsername(String username) {
+        try {
+            return create()
+                    .select(ACCOUNT.fields())
+                    .from(ACCOUNT)
+                    .join(CREDENTIAL)
+                    .on(CREDENTIAL.ACCOUNT_ID.eq(ACCOUNT.ID))
+                    .where(
+                            ACCOUNT.STATE.eq(CommonStatesConstants.ACTIVE)
+                                    .and(CREDENTIAL.STATE.eq(CommonStatesConstants.ACTIVE))
+                                    .and(CREDENTIAL.PUBLIC_VALUE.eq(username)))
+                    .and(CREDENTIAL.KIND.eq("password"))
+                    .fetchOneInto(AccountRecord.class);
+        } catch (InvalidResultException e) {
+            throw new ClientVisibleException(ResponseCodes.CONFLICT, "MultipleOfUsername");
```

Comment:
@hibooboo2 Thats a tricky one. Applying unique key would fail when duplicates are present. And in this case you can't run pre-upgrade cleanup on user's behalf as duplicates weren't introduced by some bug in our code where you would know which data is safe to remove; we've just let user create duplicated values, and now don't know which one of them is valid. 

@vincent99 @ibuildthecloud any tips on that part?